src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs

```diff
@@ -715,14 +715,19 @@ public async Task<bool> WriteConst(LiteralExpressionSyntax constValue, MonoSDBHe
             return false;
         }
 
-        public async Task<bool> WriteJsonValue(JObject objValue, MonoSDBHelper SdbHelper, CancellationToken token)
+        public async Task<bool> WriteJsonValue(JObject objValue, MonoSDBHelper SdbHelper, ElementType? expectedType, CancellationToken token)
         {
             switch (objValue["type"].Value<string>())
             {
                 case "number":
                 {
-                    // FixMe: what if the number is not int but single/double?
-                    Write(ElementType.I4, objValue["value"].Value<int>());
+                    var expected = expectedType is not null ? expectedType.Value : ElementType.I4;
+                    if (expected == ElementType.R4)
+                        Write(expected, objValue["value"].Value<float>());
+                    else if (expected == ElementType.R8)
+                        Write(expected, objValue["value"].Value<double>());
+                    else
+                        Write(expected, objValue["value"].Value<int>());
```

Comment:
If we cast then we probably need all cases:
- [ ] ElementType.I2,
- [ ] ElementType.I4,
- [ ] ElementType.I8,
- [x] ElementType.R4,
- [x] ElementType.R8,
- [ ] ElementType.U2,
- [ ] ElementType.U4,
- [ ] ElementType.U8.
Or..let's test if adding tests with them won't throw here, maybe cast to int to all other cases is enough.